### ISSUE TITLE
Different outputs in nn.MultiheadAttention with exactly the same weights and inputs during training mode

### ISSUE THREAD

**Issue by v-qjqs** [pytorch/pytorch#46697]
## ❓ Questions and Help
Hi I'm newer to transformer and I have questions to nn.MultiheadAttention.
I'm implementing DETR in the mmdet framework following the official DETR repo. I found nn.MultiheadAttention output different 'attn_output' and 'attn_output_weights' during **training** mode  between two independent experiments, even though I loaded exactly the same learnable weights and inputs (including query, key, value, attn_mask and key_padding_mask), also the MultiheadAttention is initialized with the same setting.
While during the **inference** mode, it's normal and expected that the outputs is exactly the same. 
I wonder whether this is normal during the training mode? Or is there another (randomly) factors that affects the outputs of MultiheadAttention during the training mode?
Any helps or suggestions is appreciated. Thanks very much.

### Please note that this issue tracker is not a help form and this issue will be closed.

We have a set of [listed resources available on the website](https://pytorch.org/resources). Our primary means of support is our discussion forum:

- [Discussion Forum](https://discuss.pytorch.org/)

cc @zhangguanheng66

**Comment by zhangguanheng66** [comment]
Do you see this difference during inference? It would be great if you could put together a simple code snippet so I could reproduce the difference.

Did you change `dropout` to non-zero? `dropout` introduces some randomness.

**Comment by v-qjqs** [comment]
Hi @zhangguanheng66, thanks for your comments. There is no difference during inference.
Yeah I use dropout with non-zero probability 0.1, I guess this might be the reason and I will check it.
Thanks.
